js/registration.js

```diff
@@ -0,0 +1,117 @@
+
+    errorDisplay.innerText = message;
+    formControl.classList.add('error');
+    formControl.classList.remove('success')
+};
+
+const setSuccess = element => {
+    const formControl = element.parentElement;
+    const errorDisplay = formControl.querySelector('small');
+
+    errorDisplay.innerText = '';
+    formControl.classList.add('success');
+    formControl.classList.remove('error');
+};
+
+const isValidUsername = username => {
+   const str = /^[A-Za-z0-9 ]*$/;
+   return str.test(username);
+}
+
+const isValidEmail = email => {
+    const regExEmail = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+    return regExEmail.test(String(email).toLowerCase());
+};
+
+const validateInputs = () => {
+    const usernameValue = username.value.trim();
+    const passwordValue = password.value.trim();
+    const emailValue = email.value.trim();
+    const email2Value = email2.value.trim();
+
+    if(usernameValue === '') {
+        setError(username, 'Nazwa użytkownika jest wymagana');
+    } else if (!isValidUsername(usernameValue)) {
+        setError(username, 'Nazwa użytkownika może składać się z liter i/lub liczb');
+    } else if (usernameValue.length < 6) {
+        setError(username, 'Nazwa użytkownika musi mieć conajmniej 6 znaków' );
+    } else if (usernameValue.length > 16 ) {
+        setError(username, 'Nazwa użytkownika może mieć maksymalnie 16 znaków' );
+    } else {
+        setSuccess(username);
+    }
+
+    if(passwordValue === '') {
+        setError(password, 'Hasło jest wymagane');
+    } else if (passwordValue.length < 6 ) {
+        setError(password, 'Hasło musi mieć conajmniej 6 znaków');
+    } else {
+        setSuccess(password);
+    }
+
+    if(emailValue === '') {
+        setError(email, 'Adres email jest wymagany');
+    } else if (!isValidEmail(emailValue)) {
+        setError(email, 'Podaj poprawny adres email');
+    } else {
+        setSuccess(email);
+    }
+
+    if(email2Value === '') {
+        setError(email2, 'Potwierdź adres email');
+    } else if (email2Value !== emailValue) {
+        setError(email2, 'Adresy email różnią się');
+    } else {
+        setSuccess(email2);
+    }
+};
+
+function saveData(){
+    let username = document.getElementById('username').value;
+    let password = document.getElementById('password').value;
+    let email = document.getElementById('email').value;
+    let email2 = document.getElementById('email2').value;
+
+    let user_records = new Array();
+    user_records = JSON.parse(localStorage.getItem('users'))?JSON.parse(localStorage.getItem('users')):[]
+
+    if (user_records.some((v)=>{return v.email == email})){
+        setError(email, 'Podany email już istnieje w bazie');
+    } else {user_records.push({
+      'username': username,
+      'password': password,
+      'email': email,
+      'email2': email2
+    })
```

Comment:
`push` mutuje array (doesitmutate.xyz/). Chcesz unikac mutacji. Wobec tego:
```js
const newUser = {
      'username': username,
      'password': password,
      'email': email,
      'email2': email2
    }
user_records = [...user_records, newUser] 
```

mozesz tez smialo zmienic `user_records` na po prostu `users`.

Reply:
Zrobiłam tak jak radzisz, ale teraz nie dodaje mi do tablicy danych - tzn. zapisuje biezace dane, a następne dane jakby znikały - mam tylko nowe dane - w tablicy tylko jeden uzytkownik...

Reply:
Jeszcze takie pytanie czemu tu powinnan unikac mutacji ? Chce dodawac uzytkowników do mojej bazy, wiec czemu to jest złe?